File: html/old/backup/sacor2_2007/cards.py
# Solution to the problem "Cards" from the 2007 South African Computer Olympiad
#
# Timothy Stranex (timothy.stranex@gmail.com)
# 2007-04-25

print 'Enter the numbers of the cards:',
cards = map(int, raw_input().split(' '))
cards.sort()

njokers = 0
for c in cards:
    if (c == 0):
        njokers += 1

best = 0
if (njokers == len(cards)):
    best = njokers

# there is always a longest run starting on a non-joker card if there is at
# least one non-joker card
for i in range(njokers, len(cards)):
    n = 1
    nj = 0
    
    while 1:
        if ((i+n-nj < len(cards)) and (cards[i+n-nj] == cards[i]+n)):
            n += 1
        elif (nj < njokers):
            n += 1
            nj += 1
        else:
            break
    
    best = max(best, n)

print 'The longest run has length %d.' % best

File: html/old/backup/sacor2_2007/cards.cpp
/*
 * Solution to the problem "Cards" from the
 * 2007 South African Computer Olympiad
 *
 * Timothy Stranex <timothy.stranex@gmail.com>
 * 2007-04-25
 */

#include <algorithm>
#include <vector>
#include <iostream>
#include <sstream>
using namespace std;

int main(int argc, char *argv[])
{
    vector<int> cards;

    cout << "Enter the numbers of the cards: ";
    string line;
    getline(cin, line);

    istringstream is(line);
    {
        int n;
        while (is >> n) {
            cards.push_back(n);
        }
    }
    sort(cards.begin(), cards.end());
    int N = cards.size();

    int njokers = 0;
    for (int i=0; i<N; i++) {
        if (cards[i] == 0) {
            njokers++;
        }
    }

    int best = 0;
    if (njokers == N) {
        best = N;
    }

    // there is always a longest run starting on a non-joker card if there is
    // at least one non-joker card
    for (int i=njokers; i<N; i++) {
        int n = 1;
        int nj = 0;

        while(true) {
            if ((i+n-nj < N) && (cards[i+n-nj] == cards[i]+n)) {
                n++;
            } else if (nj < njokers) {
                n++;
                nj++;
            } else {
                break;
            }
        }

        best = max(best, n);
    }

    cout << "The longest run has length " << best << "." << endl;

    return 0;
}
